File: leetcode/medium/1493_longestSubarray.py
from typing import List
from json import dumps

class Solution:
    def longestSubarray(self, nums: List[int]) -> int:
        zero = -1
        length = 0

        start = 0
        for i in range(len(nums)):
            if nums[i] == 0:
                if zero == -1:
                    zero = i
                else:
                    length = max(length, i - start - 1)
                    start = zero + 1
                    zero = i
            if i == len(nums) - 1:
                length = max(length, i - start)

        return length


def test ():
    params = [
        {
            'input': [1,1,0,1],
            'output': 3,
        },
        {
            'input': [0,1,1,1,0,1,1,0,1],
            'output': 5,
        },
        {
            'input': [1,1,1],
            'output': 2,
        },
    ]
    solution = Solution()

    for param in params:
        result = solution.longestSubarray(param['input'])
        print(
            'SUCCESS' if result == param['output'] else 'ERROR',
            'input', param['input'],
            'output', param['output'],
            'result', result,
            '\n',
        )


if __name__ == '__main__':
    test()

File: leetcode/medium/1493_longestSubarray.ts
function longestSubarray(nums: number[]): number {
    let res: number = 0

    let left: number = 0;
    let ones: number = 0;
    let zeros: number = 0;

    for (let right: number = 0; right < nums.length; ++right) {
        while (zeros > 1) {
            if (nums[left] === 0) {
                zeros -= 1;
            } else {
                ones -= 1;
            }
            left += 1;
        }

        if (nums[right] === 0) {
            zeros += 1;
        } else {
            ones += 1;
        }

        const o: number = zeros > 0 ? ones : ones - 1;

        res = Math.max(res, o);
    }

    return res;
};

const test = () => {
    const params = [

        {
            input: {
                nums: [1,1,1]
            },
            output: 2,
        },
        {
            input: {
                nums: [1,1,0,1],
            },
            output: 3,
        },
        {
            input: {
                nums: [0,1,1,1,0,1,1,0,1]
            },
            output: 5,
        },
    ];

    params.forEach(({input, output}) => {
        const { nums } = input;

        const result = longestSubarray(nums);

        console.log(
            JSON.stringify(result) === JSON.stringify(output) ? 'SUCCESS ' : 'ERROR ',
            'input', JSON.stringify(input),
            'output', JSON.stringify(output),
            'result', JSON.stringify(result),
        );
    });
};

test();
